File: NextcloudTalk/Settings/NCUserDefaults.m
//
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "NCUserDefaults.h"

#import "NCKeyChainController.h"

@implementation NCUserDefaults

NSString * const kNCPreferredCameraFlashMode    = @"ncPreferredCameraFlashMode";
NSString * const kNCBackgroundBlurEnabled       = @"ncBackgroundBlurEnabled";
NSString * const kNCIncludeCallsInRecents       = @"ncIncludeCallsInRecents";
NSString * const kNCPreferredCallViewMode       = @"ncPreferredCallViewMode";
NSString * const kNCSpeakerViewStripeHidden     = @"ncSpeakerViewStripeHidden";

+ (void)setPreferredCameraFlashMode:(NSInteger)flashMode
{
    [[NSUserDefaults standardUserDefaults] setObject:@(flashMode) forKey:kNCPreferredCameraFlashMode];
}

+ (NSInteger)preferredCameraFlashMode
{
    return [[[NSUserDefaults standardUserDefaults] objectForKey:kNCPreferredCameraFlashMode] integerValue];
}

+ (void)setBackgroundBlurEnabled:(BOOL)enabled
{
    [[NSUserDefaults standardUserDefaults] setObject:@(enabled) forKey:kNCBackgroundBlurEnabled];
}

+ (BOOL)backgroundBlurEnabled
{
    return [[[NSUserDefaults standardUserDefaults] objectForKey:kNCBackgroundBlurEnabled] boolValue];
}

+ (void)setIncludeCallsInRecentsEnabled:(BOOL)enabled
{
    [[NSUserDefaults standardUserDefaults] setObject:@(enabled) forKey:kNCIncludeCallsInRecents];
}

+ (BOOL)includeCallsInRecents
{
    id includeCallsInRecentsObject = [[NSUserDefaults standardUserDefaults] objectForKey:kNCIncludeCallsInRecents];
    if (includeCallsInRecentsObject == nil) {
        [self setIncludeCallsInRecentsEnabled:YES];
        return YES;
    }

    return [includeCallsInRecentsObject boolValue];
}

+ (void)setPreferredCallViewMode:(NSString *)mode
{
    [[NSUserDefaults standardUserDefaults] setObject:mode forKey:kNCPreferredCallViewMode];
}

+ (NSString * _Nullable)preferredCallViewMode
{
    return [[NSUserDefaults standardUserDefaults] stringForKey:kNCPreferredCallViewMode];
}

+ (void)setSpeakerViewStripeHidden:(BOOL)hidden
{
    [[NSUserDefaults standardUserDefaults] setObject:@(hidden) forKey:kNCSpeakerViewStripeHidden];
}

+ (BOOL)speakerViewStripeHidden
{
    return [[[NSUserDefaults standardUserDefaults] objectForKey:kNCSpeakerViewStripeHidden] boolValue];
}

@end

File: NextcloudTalk/Settings/NCUserDefaults.swift
//
// SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

public class NCUserDefaults: NSObject {

    private static let preferredCameraFlashModeKey = "ncPreferredCameraFlashMode"
    private static let backgroundBlurEnabledKey = "ncBackgroundBlurEnabled"
    private static let includeCallsInRecentsKey = "ncIncludeCallsInRecents"
    private static let preferredCallViewModeKey = "ncPreferredCallViewMode"
    private static let speakerViewStripeHiddenKey = "ncSpeakerViewStripeHidden"

    public static func setPreferredCameraFlashMode(_ flashMode: Int) {
        UserDefaults.standard.set(flashMode, forKey: preferredCameraFlashModeKey)
    }

    public static func preferredCameraFlashMode() -> Int {
        return UserDefaults.standard.integer(forKey: preferredCameraFlashModeKey)
    }

    public static func setBackgroundBlurEnabled(_ enabled: Bool) {
        UserDefaults.standard.set(enabled, forKey: backgroundBlurEnabledKey)
    }

    public static func backgroundBlurEnabled() -> Bool {
        return UserDefaults.standard.bool(forKey: backgroundBlurEnabledKey)
    }

    public static func setIncludeCallsInRecentsEnabled(_ enabled: Bool) {
        UserDefaults.standard.set(enabled, forKey: includeCallsInRecentsKey)
    }

    public static func includeCallsInRecents() -> Bool {
        // Defaults to enabled, and persists that default on first read
        guard UserDefaults.standard.object(forKey: includeCallsInRecentsKey) != nil else {
            self.setIncludeCallsInRecentsEnabled(true)

            return true
        }

        return UserDefaults.standard.bool(forKey: includeCallsInRecentsKey)
    }

    public static func setPreferredCallViewMode(_ mode: String) {
        UserDefaults.standard.set(mode, forKey: preferredCallViewModeKey)
    }

    public static func preferredCallViewMode() -> String? {
        return UserDefaults.standard.string(forKey: preferredCallViewModeKey)
    }

    public static func setSpeakerViewStripeHidden(_ hidden: Bool) {
        UserDefaults.standard.set(hidden, forKey: speakerViewStripeHiddenKey)
    }

    public static func speakerViewStripeHidden() -> Bool {
        return UserDefaults.standard.bool(forKey: speakerViewStripeHiddenKey)
    }
}
